src/coreclr/jit/assertionprop.cpp

```diff
@@ -3177,6 +3177,67 @@ GenTree* Compiler::optVNConstantPropOnTree(BasicBlock* block, GenTree* tree)
 
     if (conValTree != nullptr)
     {
+        if (tree->OperIs(GT_LCL_VAR) && conValTree->OperIs(GT_CNS_VEC, GT_CNS_DBL))
+        {
+            // A simple heuristic: If the constant is defined outside of a loop (not far from its head)
+            // and is used inside it - don't propagate.
+
+            bool keepPropagating = false;
+
+            if (conValTree->OperIs(GT_CNS_VEC))
+            {
+                if (conValTree->IsVectorZero() || conValTree->IsVectorAllBitsSet())
+                {
+                    // These are cheap to materialize
+                    keepPropagating = true;
+                }
+            }
+            else if (conValTree->OperIs(GT_CNS_DBL))
+            {
+                if (conValTree->IsFloatPositiveZero())
+                {
+                    // This is cheap to materialize
+                    keepPropagating = true;
+                }
+#if defined(TARGET_ARM64)
+                if (emitter::canEncodeFloatImm8(conValTree->AsDblCon()->gtDconVal))
+                {
+                    // Such floats are likely immable on arm64
+                    keepPropagating = true;
+                }
+#endif
+            }
```

Comment:
Did you end up not keeping the path that handles `GT_CNS_INT` or is that being left to a future PR since what's most impactful (at least for the regression) is CNS_VEC/CNS_DBL?